src/hw/hw_mngr.cpp

```diff
@@ -357,3 +458,179 @@ void HWManager::setFanSpeedDelayed()
         setFanSpeed();
     });
 }
+
+void HWManager::detectFansDelayed()
+{
+    static boost::asio::deadline_timer tmr(io);
+
+    //  if we has expired attempts, then stop fan detection and work in normal
+    //  state
+    if (++numErrorAttempts > maxErrorAttempts)
+    {
+        log<level::ERR>("FAN Detection",
+                        entry("WHAT=reached maxErrorAttempts in state=%d",
+                              int(detectFansState)));
+        detectFansState = NormalModeFans;
+        return;
+    }
+
+    // this implicitly cancels the timer
+    tmr.expires_from_now(boost::posix_time::seconds(5));
+    tmr.async_wait([&](const boost::system::error_code& err) {
+        if (err == boost::asio::error::operation_aborted)
+        {
+            /* we were canceled*/
+            return;
+        }
+        else if (err)
+        {
+            log<level::ERR>("Timer error",
+                            entry("WHAT=%s", err.message().c_str()));
+            return;
+        }
+
+        detectFansByRpmSpeed();
+    });
+}
+
+void HWManager::detectFansByRpmSpeed()
+{
+    log<level::DEBUG>("FAN: process detection by maximum RPM",
+                      entry("SYSTEM=current state=%d", int(detectFansState)));
+
+    if (!processSystemFans())
+    {
+        detectFansState = GetCurrentFansPwm;
+        detectFansDelayed();
+        return;
+    }
+
+    switch (detectFansState)
+    {
+        case GetCurrentFansPwm:
+            log<level::INFO>("FAN: set all fans PWM to 100%");
+            detectFansState = SetMaximumFansPwm;
+            detectFansByRpmSpeed();
+            break;
+
+        case SetMaximumFansPwm:
+            log<level::INFO>("FAN: set all fans PWM to default values");
+            detectFansState = RestoreFansPwm;
+            detectFansDelayed();
+            break;
+
+        case RestoreFansPwm:
+            updateFanPartNumber();
+            detectFansState = NormalModeFans;
+            log<level::INFO>("FAN: detection has completed");
+            break;
+
+        default:
+            break;
+    }
+}
+
+bool HWManager::processSystemFans()
+{
+    if (!fans.size())
```

Comment:
`fans.empty()` would be more human readable.

Reply:
fixed

Reply:
ops, sorry. fixed again.

Reply:
ops, sorry. fixed again.